scarlet2/plot.py

```diff
@@ -0,0 +1,883 @@
+                break
+        H_ = H
+    return H/(i+1) 
+
+def hallucination_score(scene, obs, src_num):
+    src = scene.sources[src_num]
+    center = np.array(src.morphology.bbox.center)[::-1]
+    morph = src.morphology.data
+    f = lambda morph: neural_grad(morph, src)
+    jit_hvp_x2 = jit(lambda z: hvp_grad(f, (morph,), (z,)))
+    hvp_nn = hvp_rad(jit_hvp_x2, morph.shape)
+    hvp_nn = np.array(hvp_nn)
+
+    model_scene = scene()
+    morph = model_scene[src_num] 
+    spectrum = jnp.array((1,))
+    data = obs.data
+    weights = obs.weights
+
+    # jit the HVP for this loss and this morph model
+    f = lambda morph: log_like(morph, spectrum, data, weights)
+    jit_hvp_x = jit(lambda z: hvp(f, (morph,), (z,)))
+    hvp_ll = hvp_rad(jit_hvp_x, morph.shape)
+    
+    box_size = hvp_nn.shape[1] 
+    # Cut out the square box
+    hvp_ll_cut = cut_square_box(hvp_ll, center, box_size)
+    hallucination = -hvp_nn + hvp_ll_cut
+    return hallucination
+
+def sources(
+    scene,
+    observation=None,
+    norm=None,
+    channel_map=None,
+    show_model=True,
+    show_hallucination=False,
+    show_observed=False,
+    show_rendered=False,
+    show_spectrum=True,
+    figsize=None,
+    model_mask=None,
+    add_markers=True,
+    add_boxes=False,
+):
+
+    sources = scene.sources
+    n_sources = len(sources)
+    panels = sum((show_model, show_hallucination,show_observed, show_rendered, show_spectrum))
+    f_size = 22
+
+    if figsize is None:
+        figsize = (panel_size * panels, panel_size * n_sources)
+
+    fig, ax = plt.subplots(n_sources, panels, figsize=figsize, squeeze=False)
+
+    marker_kwargs = {"mew": 1, "ms": 10}
+    box_kwargs = {"facecolor": "none", "edgecolor": "w", "lw": 0.5}
+
+    skipped = 0
+    for k, src in enumerate(sources):
+        
+        if hasattr(src.morphology.bbox, "center") and src.morphology.bbox.center is not None:
+            center = np.array(src.morphology.bbox.center)[::-1]
+        else:
+            center = None
+        
+        start, stop = src.morphology.bbox.start[-2:][::-1], src.morphology.bbox.stop[-2:][::-1]
+        points = (start, (start[0], stop[1]), stop, (stop[0], start[1]))
+        box_coords = [
+            p for p in points
+        ]
+    
+        # model in its bbox
+        panel = 0
+        model = src()
+        if show_model:
+            # Show the unrendered model in it's bbox
+            extent = get_extent(src.morphology.bbox)
+            ax[k-skipped][panel].imshow(
+                img_to_rgb(model, norm=norm, channel_map=channel_map, mask=model_mask),
+                extent=extent,
+                origin="lower",
+            )
+            ax[k-skipped][panel].set_title("Model Source {}".format(k), fontsize=f_size)
+            if center is not None and add_markers:
+                ax[k-skipped][panel].plot(*center, "wx", **marker_kwargs)
+            panel += 1
+            
+        if show_hallucination:
```

Comment:
does this work if the morphology does not have a prior? Also, does this assume that the prior is a score models?

Reply:
Must have a prior, does not need to be a score model, just the log_prob() call must return the gradients for this method to work.

Reply:
Added an assert statement asserting the morphology prior must no be `None`